### PR TITLE
2.x Add concatMapCompletable() to Observable

### PR DESCRIPTION
**Pull request by dweebo**
Add `concatMapCompletable()` to `Observable` as discussed in #4853
I didn't think it made sense in other reactive types.

Code is mostly a copy of `ObservableConcatMap`.  Let me know if there is a better style of code to base this off instead.  It also does not have the option to delay errors as concatMap() does, not sure if that is needed.

### CODE REVIEW

File: src/main/java/io/reactivex/Observable.java [hunk 1]
```diff
@@ -6252,6 +6252,46 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+     * order and waits until the upstream and all CompletableSources complete.
+     * <dl>
+     *  <dt><b>Scheduler:</b></dt>
+     *  <dd>{@code concatMapCompletable} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param mapper
+     *            a function that, when applied to an item emitted by the source ObservableSource, returns a CompletableSource
+     * @return a Completable that signals {@code onComplete} when the upstream and all CompletableSources complete
+     */
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: Please add `@since 2.1.6 - experimental`

File: src/main/java/io/reactivex/Observable.java [hunk 2]
```diff
@@ -6252,6 +6252,46 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+    @CheckReturnValue
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final Completable concatMapCompletable(Function<? super T, ? extends CompletableSource> mapper) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: Please add `@Experimental`.

File: src/main/java/io/reactivex/Observable.java [hunk 3]
```diff
@@ -6252,6 +6252,46 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+     *  <dt><b>Scheduler:</b></dt>
+     *  <dd>{@code concatMapCompletable} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param mapper
+     *            a function that, when applied to an item emitted by the source ObservableSource, returns a CompletableSource
+     * @param prefetch
+     *            the number of elements to prefetch from the current Observable
+     * @return a Completable that signals {@code onComplete} when the upstream and all CompletableSources complete
+     */
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: Please add `@since 2.1.6`.

File: src/main/java/io/reactivex/Observable.java [hunk 4]
```diff
@@ -6252,6 +6252,46 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+    @CheckReturnValue
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final Completable concatMapCompletable(Function<? super T, ? extends CompletableSource> mapper, int prefetch) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: Please add `@Experimental`.

File: src/main/java/io/reactivex/Observable.java [hunk 5]
```diff
@@ -6252,6 +6252,46 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+     * Maps each element of the upstream Observable into CompletableSources, subscribes to them one at a time in
+     * order and waits until the upstream and all CompletableSources complete.
+     * <dl>
+     *  <dt><b>Scheduler:</b></dt>
+     *  <dd>{@code concatMapCompletable} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param mapper
+     *            a function that, when applied to an item emitted by the source ObservableSource, returns a CompletableSource
+     * @param prefetch
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Observable.java]: A `capacityHint` would capture the underlying behavior better: "the number of upstream items expected to be buffered until the current CompletableSource, mapped from the current item, completes."

**Review by akarnokd — COMMENTED** [review]
Codewise it looks okay. /cc @artem-zinnatullin, @vanniktech & @davidmoten who was also interested in this.

File: src/main/java/io/reactivex/Observable.java [hunk 6]
```diff
@@ -6252,6 +6252,52 @@ public final void blockingSubscribe(Observer<? super T> subscriber) {
...
+     * @param mapper
+     *            a function that, when applied to an item emitted by the source ObservableSource, returns a CompletableSource
+     * @return a Completable that signals {@code onComplete} when the upstream and all CompletableSources complete
+     * @since 2.1.6 - experimental
+     */
+    @Experimental
+    @CheckReturnValue
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final Completable concatMapCompletable(Function<? super T, ? extends CompletableSource> mapper) {
+        return concatMapCompletable(mapper, 2);
```
Comment (vanniktech) [inline src/main/java/io/reactivex/Observable.java:6271]: Should not we pass 8 here since it'll get overridden to 8 in the implementation?

**Review by vanniktech — APPROVED** [review]
Should we also include a diagram?

**Comment by akarnokd** [comment]
Added to #5319 for the diagram.

File: src/main/java/io/reactivex/Observable.java (hunk 6 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/Observable.java:6271]: Leave it for now.

**Pull request merged**